### PR TITLE
[stablehlo] Remove potential double-free in ReorderBroadcastInDimOpAndElementwiseOp

### PR DESCRIPTION
**Pull request by nbondoux**
This fix a double free that occurs when the two Broadcast operations are the same operation.

### CODE REVIEW

**Comment by hanhanW** [comment]
When does it happen? Can you add a test?

**Comment by nbondoux** [comment]
> When does it happen? Can you add a test?

It happens in the case where you have
%1= Broadcast(....)
%2=ElementWiseOp(%1,%1, ....)
the Broadcast op will be erased two times. Concretely, it happened while I was testing iree against a jax model, using the PJRT plugin. I can add a test later.

File: compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/StableHLOToStableHLO.cpp [hunk 1]
```diff
@@ -1186,12 +1186,16 @@ struct ReorderBroadcastInDimOpAndElementwiseOp final
...
     for (auto bcastOp : bcastOps) {
       if (bcastOp.getOperation()->use_empty()) {
-        rewriter.eraseOp(bcastOp);
+        opsToErase.insert(bcastOp.getOperation());
       }
     }
 
+    for(auto opToErase: opsToErase) {
+      rewriter.eraseOp(opToErase);
+    }
```
Comment (hanhanW) [inline compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/StableHLOToStableHLO.cpp]: Maybe we can just remove these, and rely on cse to delete the ops?

**Review by hanhanW — COMMENTED** [review]
I see, thanks

**Comment by benvanik** [comment]
Rebased and added a test in #17394.

**Pull request closed without merging**